### PR TITLE
#3: Image Component Enhanced

### LINKED ISSUE #3: Enhance Image Vue component
## Description

Enhance https://github.com/ripe-tech/ripe-commons-pluginus/issues/135 with...

... some props: format, size, crop

Unit testing and documentation (according to internal vue.js standards) is required.

### PR DESCRIPTION
**Pull request by BeeMargarida**
| - | - |
| --- | --- |
| Issue | Closes #3 |
| Dependencies | https://github.com/ripe-tech/ripe-sdk/pull/192 |
| Decisions | Image component with new props (crop, format) and initials logic (initialsBuilder, initialsGroup, showInitials and state). The state contains information about the initials and engraving. |
| Animated GIF | image-personalizationimage_enh|

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -181,9 +183,44 @@ The customization of the model can also be provided, with the prop parts:
...
+    v-bind:initials-group="'main'"
+    v-bind:state="state"
+/>
+```
+
+![Image with Personalization](res/images/image-personalization.png)
+
+It is also possible to provide a `initialsBuilder` function, that allows for a more custom logic in the translation from initials and engraving into initials and profiles.
+
+There can be more than one image using the same instance of Ripe SDK:
```
Comment (gcandal) [inline README.md:253]: Maybe this should be under a different section, right now it's inside `Image` which I don't think it's what you intended right?

Comment (BeeMargarida) [reply README.md:253]: @gcandal The part with the multiple SDKs, or the one related to the `initialsBuilder`?

Comment (BeeMargarida) [reply README.md:253]: Both of them are related to the image

**Comment by gcandal** [comment]
@BeeMargarida it seems that changing `crop` in the Storybook doesn't produce any changes. Can you please take a look at what's going on?

**Comment by gcandal** [comment]
Also, I guess there could be unit tests for this: instantiate an image, change stuff in the props, check that the image URL changes accordingly.

**Comment by BeeMargarida** [comment]
> Also, I guess there could be unit tests for this: instantiate an image, change stuff in the props, check that the image URL changes accordingly.

@gcandal Regarding the testing, its a bit more complicated than that. The setup of Ripe SDK is async, however it happens when the component is mounted, so there is no way in the test for waiting for the configuration to end. Its also very complicated by mocking the SDK because of the bindImage method, which must return an Image that implements Observable methods (bind, unbind, etc), which would involve an extensive mock

**Comment by BeeMargarida** [comment]
@gcandal Regarding the tests, even by using stubs and mocks, it is not possible. When mounting the component during the tests, the mounting returns before the async methods finish. It makes it impossible to tests behaviour related to the image, which only becomes available after the ripe sdk configuration ends.

File: vue/components/organisms/ripe-image/ripe-image.vue [hunk 2]
```diff
@@ -248,7 +234,8 @@ export const RipeImage = {
             try {
                 await this.ripeData.config(this.brand, this.model, {
                     version: this.version,
-                    parts: this.partsData
+                    parts: this.parts,
+                    safe: true
```
Comment (gcandal) [inline vue/components/organisms/ripe-image/ripe-image.vue:238]: 💪

**Review by gcandal — APPROVED** [review]

**Review by gcandal — APPROVED** [review]

**Review by joamag — APPROVED** [review]

**Pull request merged**